File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.cc
#include <algorithm>
#include <iostream>
#include <string>
#include <vector>
using namespace std;

int g[25][25];
bool v[25][25];
int dx[4] = {-1, 1, 0, 0};
int dy[4] = {0, 0, -1, 1};
int n;

int dfs(int x, int y) {
  int cnt = 1;
  v[x][y] = true;

  for (int i = 0; i < 4; i++) {
    int nx = x + dx[i], ny = y + dy[i];
    if (nx >= 0 && ny >= 0 && nx < n && ny < n && !v[nx][ny] &&
        g[nx][ny] == 1) {
      cnt += dfs(nx, ny);
    }
  }
  return cnt;
}

int main() {
  cin >> n;
  cin.ignore();

  for (int i = 0; i < n; i++) {
    string line;
    getline(cin, line);
    for (int j = 0; j < n; j++) {
      g[i][j] = line[j] - '0';  // 아스키 자동 변환
    }
  }

  vector<int> danzis;
  int cnt = 0;

  for (int i = 0; i < n; i++) {
    for (int j = 0; j < n; j++) {
      if (!v[i][j] && g[i][j] == 1) {
        int size = dfs(i, j);
        danzis.push_back(size);
        cnt++;
      }
    }
  }

  cout << cnt << "\n";
  sort(danzis.begin(), danzis.end());

  for (int i = 0; i < danzis.size(); i++) {
    cout << danzis[i] << "\n";
  }

  return 0;
}

File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.java
import java.io.BufferedReader;
import java.io.FileInputStream;
import java.io.InputStreamReader;
import java.util.Arrays;


public class Main {
	static int N;
	static int no = 1;
	static int[][] map;
	static int[] dx = {-1, 1, 0, 0};
	static int[] dy = {0, 0, -1, 1};

	public static void main(String[] args) throws Exception {
		//System.setIn(new FileInputStream("4. Algorithm /Algo/src/boj/silver/input.txt"));
		BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
		N = Integer.parseInt(in.readLine());
		map = new int[N][N];
		for (int i = 0; i < N; i++) {
			String line = in.readLine();
			for (int j = 0; j < N; j++) {
				map[i][j] = line.charAt(j) - '0';
			}
		}
		for (int i = 0; i < N; i++) {
			for (int j = 0; j < N; j++) {
				if (map[i][j] == 1) {
					no++;
					dfs(i, j, no);
				}
			}
		}
		--no;
		System.out.println(no);
		int[] count = new int[no + 2];
		for (int i = 0; i < N; i++) {
			for (int j = 0; j < N; j++) {
				if (map[i][j] > 0) {
					count[map[i][j]]++;
				}
			}
		}
		Arrays.sort(count);

		for (int i = 2; i <= no + 1; i++) {
			System.out.println(count[i]);
		}
	}

	public static void dfs(int x, int y, int no) {
		map[x][y] = no;
		for (int i = 0; i < 4; i++) {
			int nx = x + dx[i];
			int ny = y + dy[i];
			if (nx >= 0 && nx < N && ny >= 0 && ny < N && map[nx][ny] == 1) {
				dfs(nx, ny, no);
			}
		}
	}
}